File: AndroidApp/shemsu-shop/app/src/main/java/com/example/data/remote/ApiService.kt
package com.example.data.remote

import com.example.data.model.*
import retrofit2.Response
import retrofit2.http.*

data class OtpRequest(val username: String)
data class OtpResponse(val success: Boolean, val message: String?, val error: String?)

data class VerifyRequest(val username: String, val otp: String)
data class VerifyResponse(val success: Boolean, val session_token: String?, val error: String?)

data class ProfileRequest(val user_id: Long, val full_name: String)

data class PlaceOrderRequest(
    val user_id: Long,
    val username: String?,
    val full_name: String?,
    val items: List<CartItemDto>,
    val payment_method: String,
    val payment_account_id: Int?,
    val confirmation: String?,
    val comment: String?
)

data class CartItemDto(
    val item: String,
    val qty: Int,
    val price: Double,
    val comment: String? = null,
    val isCustom: Boolean = false
)

data class PlaceOrderResponse(
    val success: Boolean,
    val order_group: String?,
    val total: Double?,
    val payment_method: String?,
    val status: String?,
    val error: String?
)

data class DebtTotalResponse(val active_total: Double)
data class DebtPayRequest(
    val username: String,
    val user_id: Long,
    val amount: Double,
    val payment_account_id: Int,
    val confirmation: String
)
data class GenericSuccessResponse(val success: Boolean, val message: String? = null, val error: String? = null)
data class DebtAllowResponse(val allowed: Boolean)
data class SessionCheckResponse(val user_id: Long, val username: String, val full_name: String)

interface ApiService {
    @POST("api/auth/request-otp")
    suspend fun requestOtp(@Body request: OtpRequest): Response<OtpResponse>

    @POST("api/auth/verify-otp")
    suspend fun verifyOtp(@Body request: VerifyRequest): Response<VerifyResponse>

    @GET("api/auth/session")
    suspend fun checkSession(@Query("token") token: String): Response<SessionCheckResponse>

    @GET("api/user/profile")
    suspend fun getProfile(@Query("user_id") userId: Long): Response<User>

    @PUT("api/user/profile")
    suspend fun updateProfile(@Body request: ProfileRequest): Response<GenericSuccessResponse>

    @GET("api/menu")
    suspend fun getMenu(): Response<List<MenuItem>>

    @POST("api/orders")
    suspend fun placeOrder(@Body request: PlaceOrderRequest): Response<PlaceOrderResponse>

    @GET("api/orders")
    suspend fun getOrders(@Query("user_id") userId: Long): Response<List<OrderItemRemote>>

    @DELETE("api/orders/{order_group}")
    suspend fun cancelOrder(@Path("order_group") orderGroup: String): Response<GenericSuccessResponse>

    @GET("api/debts")
    suspend fun getDebts(@Query("username") username: String): Response<List<Debt>>

    @GET("api/debts/active-total")
    suspend fun getDebtActiveTotal(@Query("username") username: String): Response<DebtTotalResponse>

    @POST("api/debts/pay")
    suspend fun payDebt(@Body request: DebtPayRequest): Response<GenericSuccessResponse>

    @GET("api/payment-accounts")
    suspend fun getPaymentAccounts(): Response<List<PaymentAccount>>

    @GET("api/debt-allow-list/check")
    suspend fun checkDebtAllowed(@Query("username") username: String): Response<DebtAllowResponse>

    @POST("api/feedback")
    suspend fun submitFeedback(@Body body: HashMap<String, Any>): Response<GenericSuccessResponse>

    @GET("api/notifications")
    suspend fun getNotifications(@Query("user_id") userId: Long): Response<List<Notification>>

    @PUT("api/notifications/{id}/read")
    suspend fun markNotificationRead(@Path("id") id: Int): Response<GenericSuccessResponse>

    @POST("api/contact")
    suspend fun postContactAdmin(@Body body: HashMap<String, Any>): Response<GenericSuccessResponse>
}

data class OrderItemRemote(
    val id: Int,
    val item: String,
    val qty: Int,
    val status: String,
    val order_group: String,
    val timestamp: String
)

File: AndroidApp/shemsu-shop/flutterApp/lib/data/api_service.dart
import 'package:dio/dio.dart';
import '../models/models.dart';
import '../core/constants.dart';

class ApiService {
  final Dio _dio;

  ApiService({Dio? dio}) : _dio = dio ?? Dio(BaseOptions(
    baseUrl: AppConstants.baseUrl,
    connectTimeout: const Duration(seconds: 60),
    receiveTimeout: const Duration(seconds: 60),
    headers: {'Content-Type': 'application/json'},
  ));

  Future<OtpResponse> requestOtp(String username) async {
    final res = await _dio.post('api/auth/request-otp', data: {'username': username});
    return OtpResponse.fromJson(res.data);
  }

  Future<VerifyResponse> verifyOtp(String username, String otp) async {
    final res = await _dio.post('api/auth/verify-otp', data: {'username': username, 'otp': otp});
    return VerifyResponse.fromJson(res.data);
  }

  Future<SessionCheckResponse> checkSession(String token) async {
    final res = await _dio.get('api/auth/session', queryParameters: {'token': token});
    return SessionCheckResponse.fromJson(res.data);
  }

  Future<User> getProfile(int userId) async {
    final res = await _dio.get('api/user/profile', queryParameters: {'user_id': userId});
    return User.fromJson(res.data);
  }

  Future<GenericResponse> updateProfile(int userId, String fullName) async {
    final res = await _dio.put('api/user/profile', data: {'user_id': userId, 'full_name': fullName});
    return GenericResponse.fromJson(res.data);
  }

  Future<List<MenuItem>> getMenu() async {
    final res = await _dio.get('api/menu');
    return (res.data as List).map((i) => MenuItem.fromJson(i)).toList();
  }

  Future<PlaceOrderResponse> placeOrder({
    required int userId,
    String? username,
    String? fullName,
    required List<CartItem> items,
    required String paymentMethod,
    int? paymentAccountId,
    String? confirmation,
    String? comment,
  }) async {
    final res = await _dio.post('api/orders', data: {
      'user_id': userId,
      'username': username,
      'full_name': fullName,
      'items': items.map((e) => e.toJson()).toList(),
      'payment_method': paymentMethod,
      'payment_account_id': paymentAccountId,
      'confirmation': confirmation,
      'comment': comment,
    });
    return PlaceOrderResponse.fromJson(res.data);
  }

  Future<List<OrderGroup>> getOrders(int userId) async {
    final res = await _dio.get('api/orders', queryParameters: {'user_id': userId});
    final data = res.data;
    if (data is List) {
      return data.map((i) => OrderGroup.fromJson(i)).toList();
    }
    return [];
  }

  Future<GenericResponse> cancelOrder(String orderGroup) async {
    final res = await _dio.delete('api/orders/$orderGroup');
    return GenericResponse.fromJson(res.data);
  }

  Future<List<Debt>> getDebts(String username) async {
    final res = await _dio.get('api/debts', queryParameters: {'username': username});
    return (res.data as List).map((i) => Debt.fromJson(i)).toList();
  }

  Future<DebtTotalResponse> getDebtActiveTotal(String username) async {
    final res = await _dio.get('api/debts/active-total', queryParameters: {'username': username});
    return DebtTotalResponse.fromJson(res.data);
  }

  Future<GenericResponse> payDebt({
    required String username,
    required int userId,
    required double amount,
    required int paymentAccountId,
    required String confirmation,
  }) async {
    final res = await _dio.post('api/debts/pay', data: {
      'username': username,
      'user_id': userId,
      'amount': amount,
      'payment_account_id': paymentAccountId,
      'confirmation': confirmation,
    });
    return GenericResponse.fromJson(res.data);
  }

  Future<List<PaymentAccount>> getPaymentAccounts() async {
    final res = await _dio.get('api/payment-accounts');
    return (res.data as List).map((i) => PaymentAccount.fromJson(i)).toList();
  }

  Future<GenericResponse> checkDebtAllowed(String username) async {
    final res = await _dio.get('api/debt-allow-list/check', queryParameters: {'username': username});
    return GenericResponse.fromJson(res.data);
  }

  Future<GenericResponse> submitFeedback(int userId, int rating, String comment) async {
    final res = await _dio.post('api/feedback', data: {
      'user_id': userId,
      'rating': rating,
      'comment': comment,
    });
    return GenericResponse.fromJson(res.data);
  }

  Future<List<AppNotification>> getNotifications(int userId) async {
    final res = await _dio.get('api/notifications', queryParameters: {'user_id': userId});
    return (res.data as List).map((i) => AppNotification.fromJson(i)).toList();
  }

  Future<GenericResponse> markNotificationRead(int id) async {
    final res = await _dio.put('api/notifications/$id/read');
    return GenericResponse.fromJson(res.data);
  }

  Future<GenericResponse> contactAdmin(int userId, String username, String message) async {
    final res = await _dio.post('api/contact', data: {
      'user_id': userId,
      'username': username,
      'message': message,
    });
    return GenericResponse.fromJson(res.data);
  }
}
